share/github-backup-utils/ghe-backup-mssql

```diff
@@ -128,25 +181,89 @@ else
+          ghe_verbose "Taking a full backup instead of a diff backup because for $db the checkpoint LSN for $full_backup_file couldn't be determined"
+          backup_type="full"
+          break
+        fi
+
+        next_diff_backup_base_lsn=$(get_next_diff_backup_base_lsn "$db")
+        if [[ "$next_diff_backup_base_lsn" = "NULL" ]] || [[ "$next_diff_backup_base_lsn" == "" ]]; then
+          ghe_verbose "Taking a full backup instead of a $backup_type backup because for $db the base LSN for the next diff backup couldn't be determined"
+          backup_type="full"
+          break
+        fi
+
+        # The base of the diff backup we're about to take must exactly match the checkpoint LSN of the full backup file we have
+        if [[ "$next_diff_backup_base_lsn" -ne "$full_backup_file_checkpoint_lsn" ]]; then
+          ghe_verbose "Taking a full backup instead of a $backup_type backup because for $db the diff would have base LSN $next_diff_backup_base_lsn yet our full backup has checkpoint LSN $full_backup_file_checkpoint_lsn"
+          backup_type="full"
+          break
+        fi
+      fi
+
+      # Ensure that a transaction log backup will immediately follow the previous one
+      latest_log_backup_file=$(get_latest_backup_file "$last_mssql" "$db" "log")
+      if [[ "$latest_log_backup_file" == "" ]]; then
+        ghe_verbose "Taking a full backup instead of a $backup_type backup because for $db a previous transaction log backup wasn't found"
+        backup_type="full"
+        break
+      fi
+
+      latest_log_backup_last_lsn=$(get_backup_last_lsn "$db" "$latest_log_backup_file")
+      if [[ "$latest_log_backup_last_lsn" = "NULL" ]] || [[ "$latest_log_backup_last_lsn" == "" ]]; then
+        ghe_verbose "Taking a full backup instead of a $backup_type backup because for $db the LSN range for $latest_log_backup_file couldn't be determined"
+        backup_type="full"
+        break
+      fi
+
+      next_log_backup_starting_lsn=$(get_next_log_backup_starting_lsn "$db")
+      if [[ "$next_log_backup_starting_lsn" = "NULL" ]] || [[ "$next_log_backup_starting_lsn" == "" ]]; then
+        ghe_verbose "Taking a full backup instead of a $backup_type backup because for $db the starting LSN for the next log backup couldn't be determined"
+        backup_type="full"
+        break
+      fi
+
+      # The starting LSN of the backup we're about to take must be equal to (or before) the last LSN from the last backup,
+      # otherwise there'll be a gap and the logfiles won't be restorable
+      if [[ "$next_log_backup_starting_lsn" -gt "$latest_log_backup_last_lsn" ]]; then
+        ghe_verbose "Taking a full backup instead of a $backup_type backup because for $db a gap would exist between the last backup ending at LSN $latest_log_backup_last_lsn and next backup starting at $next_log_backup_starting_lsn"
+        backup_type="full"
+        break
+      fi
```

Comment:
Does this section need to be wrapped behind `if [ "$backup_type" == 'transaction' ]`

Reply:
Good question. The idea was to also upgrade diff backups to full (this is under `if [ "$backup_type" == 'diff' ] || [ "$backup_type" == 'transaction' ]`) since they also take a log backup, but you're quite possibly right that it isn't necessary. Let's think it through:

To restore to an arbitrary point before the diff backup, you need the full backup and log backups going from the full backup LSN up to the diff backup's LSN. The log backup chain should be intact (because of this check for log backups), _except_ possibly right before the diff backup if another mechanism took a log backup first.

To restore to a point at/after the diff backup, you need the full backup, diff backup, and only logs starting from at/before the diff backup's LSN. The log backup we take right after the diff backup will always cover its LSN, unless the unlikely race discussed elsewhere occurs.

So, I think that checking log continuity when taking a diff (+ log) backup doesn't actually increase the available restore points, but does guarantee that the set of backup files in a backup-utils snapshot can be PITR'ed to _any_ point (and "resets" the set of files to just be a new full backup).

Am I thinking about that correctly?

Reply:
> (this is under if [ "$backup_type" == 'diff' ] || [ "$backup_type" == 'transaction' ]) since they also take a log backup

I think this (i.e. diff backup also takes a log backup) answers my initial question 😃